### ISSUE TITLE
Add forget `--max-age` option

### ISSUE THREAD

**Issue by kevincox** [restic/restic#3815]
Output of `restic version`
--------------------------

restic 0.13.1 compiled with go1.17.11 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

Right now all of the `restic forget --prune` options operate relative to the most recent snapshot for each host or always keep a fixed number of items. This is generally a good idea for safety but means that old hosts will *never* be deleted.

I think it would be good to have an option such as `--max-age 2y` which would delete snapshots after 2 years even if they are the newest for the host.

It may make sense to also have options such as `--max-hourly 4w` and similar for other timeframes. This would drop hourly snapshots after 4 weeks even if they would otherwise be kept due to `--keep-hourly 6` (except for hourly snapshot which are also daily snapshots which would be kept according to `--keep-daily` and `--max-daily`). This provides a graceful way to slowly forget data about hold hosts. Note that these options would be very similar to the `--keep-within-hourly` options except they run relative to the current time, not the last snapshot.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

I am trying to configure backups so that old hosts are eventually cleaned up (by the hosts that are still running. I a looking for a config that looks something like this:

```
# Same old restic config.
--keep-last 4
--keep-daily 7
--keep-weekly 6
--keep-monthly 12

# New options
--max-hourly 7d
--max-daily 30d
--max-age 2y
```

This would not affect hosts that are regularly backing up as the "max" options are much less aggressive than the main cleanup options. However if that host stops backing up (for example it was decommissioned) it would eventually get cleaned up. With this example config the backups would be reduced to daily after 7 days, weekly after 30 days and all data would be deleted after 2 years. Notably the regular restic logic for identifying daily, weekly and so on would ensure that the latest backup would be kept for the full 2 years.

Even if all of these options are considered too complicated I would love just to see the `--max-age` option. This provides a way to stay fairly safe while not leaking data. (Or requiring dangerous manual cleanup)

**Comment by kevincox** [comment]
Maybe I am misunderstanding the documentation but these all day "relative to the latest snapshot" which makes me believe that if a group stops getting new snapshots the reference time will never advance. Do you mean that this "latest snapshot" is considering all snapshots in the repository ignoring grouping and filtering?

**Comment by metalsp0rk** [comment]
This depends on your `--group-by` choice.  By default, this is done by the host and paths.  

Restic groups all the snapshots into groups according to the group-by policy you defined, or default (host,paths).  It then applies the forget policy to _each group_.  

For example, if you have two hosts, and you choose --keep-last 7, it will keep the last 7 snapshots from _each host and path grouping_.  But if you choose `--group-by ''`, it will keep the last 7 snapshots in the repository, regardless of which host or path grouping the snapshot is for.

I run two stages of forget: 

First stage cleans up old snapshots by date:
```
restic forget --group-by '' --keep-within 4y --keep-monthly 2y --keep-weekly 6m --keep-daily 6w
```
Second stage sets the "keep-last" while grouping by host and paths:
```
restic forget --group-by host,paths' --keep-daily 24 --keep-weekly 20 --keep-monthly 24 --keep-yearly 4
```

**Comment by kevincox** [comment]
Hmm, I see. Doing a `restic forget --group-by '' --keep-within 4y` afterwords would work for the eventual cleanup. That's a clever solution and I'll add that to accomplish what I wanted from `--max-age 2y`.

But I don't think the other flags such as `--keep-within-daily 6w` would work quite as expected. Because I assume that this would only keep 1 snapshot a day for the last 6 weeks. Whereas I would still want to keep 1 snapshot per group per day for that window.

The other minor downside of doing two forget runs is that it will lock the repository for longer as it needs to do two sweeps and recompactions, but that isn't a huge deal.

**Comment by metalsp0rk** [comment]
>But I don't think the other flags such as --keep-within-daily 6w would work quite as expected. Because I assume that this would only keep 1 snapshot a day for the last 6 weeks. Whereas I would still want to keep 1 snapshot per group per day for that window.

After looking through the code, it appears you're correct.  [internal/restic/snapshot_policy.go#L282](https://github.com/restic/restic/blob/master/internal/restic/snapshot_policy.go#L282) is probably of interest to you.  I'll have to review my snapshots.  :grimacing: 

There might be some value in allowing users to define if we are basing our comparison on the most recent snapshot in the whole repository or the most recent snapshot in the snapshot group we are comparing against.  

Something like `--within 'global'` or `--within 'group'`.

**Comment by kevincox** [comment]
Yeah, that could work. It seems a lot more mind-bending than my original proposal though.

The original proposal should be pretty easy to implement. Basically if you would mark something as "daily" but it is older than `--max-daily` don't. Basically if you have `--max-daily 14d` then nothing older than 14d can be considered a daily snapshot. The same goes for the other flags other than `--max-age` which just prevents keeping anything older than that.

**Comment by metalsp0rk** [comment]
I'll whip something up next time I have a couple free hours, see if the maintainers will accept it.

**Comment by kjetilho** [comment]
See #5136 for a possible more general solution.

**Comment by MichaelEischer** [comment]
Closing in favor of https://github.com/restic/restic/issues/5193 . Adding yet another bunch of flags to the `forget` command that only minimally differs from `--keep-within*` feels like an unnecessary complication of the CLI. A `--relative now` flag would achieve the same behavior.
